app/lib/search/backend.dart

```diff
@@ -76,3 +91,102 @@ class SearchBackend {
 }
 
 String _toUrl(String package) => 'https://pub.dartlang.org/packages/$package';
+
+class SnapshotStorage {
+  final Storage storage;
+  final Bucket bucket;
+  final GZipCodec _gzip = new GZipCodec();
+
+  SnapshotStorage(this.storage, this.bucket);
+
+  Future<SearchSnapshot> fetch() async {
+    final List<BucketEntry> list = await bucket.list().toList();
+    final List<String> names = list
+        .where((entry) => entry.isObject)
+        .map((entry) => entry.name)
+        .toList();
+    if (names.isEmpty) return null;
+    // Select the first entry, because there is chance that later entries were
+    // aborted for some reason, and the most probable correct entry is the first
+    // one.
+    names.sort();
+    final String selected = names.first;
+    // reading
+    final Stream<List<int>> storageStream = bucket.read(selected);
+    final Stream<List<int>> unzippedStream = _gzip.decoder.bind(storageStream);
+    final String json = await UTF8.decoder.bind(unzippedStream).join();
```

Comment:
Small nit: This can be also chained like `bucket.read(selected).transform(_gzip.decoder).transform(UTF8.decoder).single` I think.